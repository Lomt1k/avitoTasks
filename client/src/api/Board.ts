import { z } from "zod";
import { api, validateResponse } from "./api";

export const BoardSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  taskCount: z.number(),
});

export const BoardResponseSchema = z.object({
  data: z.array(BoardSchema),
});

export type Board = z.infer<typeof BoardSchema>;

export const fetchBoards = async (): Promise<Board[]> => {
  const response = await api.get('/boards');
  validateResponse(response);
  const result = BoardResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}