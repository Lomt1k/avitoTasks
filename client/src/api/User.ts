import { z } from "zod";

export const UserSchema = z.object({
  avatarUrl: z.string(),
  description: z.string(),
  email: z.string(),
  fullName: z.string(),
  id: z.number(),
  tasksCount: z.number(),
  teamId: z.number(),
  teamName: z.string(),
});

export const AssigneeSchema = UserSchema.pick({
  avatarUrl: true,
  email: true,
  fullName: true,
  id: true,
});

export type User = z.infer<typeof UserSchema>;

export type Assignee = z.infer<typeof AssigneeSchema>;