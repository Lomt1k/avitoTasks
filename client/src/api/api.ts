import axios, { AxiosResponse } from "axios";
import { z } from "zod";

export const api = axios.create({
  baseURL: 'http://localhost:8080/api/v1',
  timeout: 1000,
  validateStatus: () => true,
});

export const ErrorMessageSchema = z.object({
  error: z.string(),
  message: z.string(),
});

export type ErrorMessage = z.infer<typeof ErrorMessageSchema>;

export const validateResponse = (response: AxiosResponse): void => {
  if (response.status === 200) {
    return;
  }

  const result = ErrorMessageSchema.safeParse(response.data);
  if (result.data) {
    const error = new Error(result.data.message);
    error.name = result.data.error;
    console.error(error);
    throw error;
  }

  const error = new Error(`${response.status} ${response.statusText}`);
  error.name = 'Network Error';
  console.error(error);
  throw error;
}