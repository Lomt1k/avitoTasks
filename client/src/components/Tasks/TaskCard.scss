.task-card {
  display: block;
  background-color: var(--color-white);
  padding: 20px;
  border-radius: 15px;
  text-align: left;
  font-family: var(--ff-main);
  border: 2px solid transparent;
  outline: none;
  cursor              : pointer;
  transition: background-color var(--transition-200);

  &:hover,
  &:focus-visible {
    background-color: #ebe8e8;
  }

  &__info {
    margin: 0 0 10px;
  }

  &__title {
    margin: 0;
    margin: 0 0 5px;
  }

  &__stickers {
    display: flex;
    flex-wrap: wrap;
    gap: 5px;
    margin: 0 0 10px;
  }

  &__sticker {
    color: var(--color-white);
    font-weight: 500;
    background-color: var(--color-grey);
    padding: 2px 8px 0;
    user-select: none;
  }

  &__desc {
    margin: 0;
  }

  &__user {
    display: flex;
    align-items: center;
    gap: 20px;
  }

  &__user-avatar {
    border-radius: 50%;
  }

  &__user-info {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 2px;
  }

  &__user-name {
    font-weight: 500;
  }

  &__user-email {
    font-size: 14px;
    color: var(--color-navy-blue);
    outline: none;
    transition: var(--transition-200);    

    &:hover,
    &:focus-visible {
      color: var(--color-green);
    }
  }
}