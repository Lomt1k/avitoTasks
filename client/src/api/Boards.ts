import { z } from "zod";
import { api, validateResponse } from "./api";

export const GeneralBoardInfoSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  taskCount: z.number(),
});

export const GeneralBoardInfoResponseSchema = z.object({
  data: z.array(GeneralBoardInfoSchema),
});

export type GeneralBoardInfo = z.infer<typeof GeneralBoardInfoSchema>;

export const fetchBoards = async (): Promise<GeneralBoardInfo[]> => {
  const response = await api.get('/boards');
  validateResponse(response);
  const result = GeneralBoardInfoResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}