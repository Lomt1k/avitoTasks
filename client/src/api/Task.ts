import { z } from "zod";
import { AssigneeSchema } from "./User";
import { api, validateResponse } from "./api";

export enum TaskStatus {
  Backlog = 'Backlog',
  InProgress = 'InProgress',
  Done = 'Done',
};

export enum TaskPriority {
  Low = 'Low',
  Medium = 'Medium',
  High = 'High',
}

export const TaskSchema = z.object({
  assignee: AssigneeSchema,
  boardId: z.number().optional(),
  boardName: z.string().optional(),
  description: z.string(),
  id: z.number(),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
  title: z.string(),
});

export const TasksResponseSchema = z.object({
  data: z.array(TaskSchema),
});

export const CreateTaskSchema = z.object({
  assigneeId: z.number(),
  boardId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
  description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
});

export const UpdateTaskSchema = z.object({
  assigneeId: z.number(),
  title: z.string()
    .nonempty('Необходимо указать название')
    .max(100, 'Должно быть не более 100 символов'),
    description: z.string()
    .nonempty('Необходимо указать описание')
    .max(500, 'Должно быть не более 500 символов'),
  priority: z.enum(Object.values(TaskPriority) as [string, ...string[]]),
  status: z.enum(Object.values(TaskStatus) as [string, ...string[]]),
});

export type Task = z.infer<typeof TaskSchema>;

export type CreateTaskData = z.infer<typeof CreateTaskSchema>;

export type UpdateTaskData = z.infer<typeof UpdateTaskSchema>;

export const fetchTasksFromBoard = async (boardId: number): Promise<Task[]> => {
  const response = await api.get(`/boards/${boardId}`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchAllTasks = async (): Promise<Task[]> => {
  const response = await api.get(`/tasks`);
  validateResponse(response);
  const result = TasksResponseSchema.safeParse(response.data);
  if (!result.data) {
    console.error(result.error);
    throw result.error;
  }
  return result.data.data;
}

export const fetchCreateTask = async (data: CreateTaskData): Promise<void> => {
  const response = await api.post('/tasks/create', data);
  validateResponse(response);
}

export const fetchUpdateTask = async (taskId: number, data: UpdateTaskData): Promise<void> => {
  const response = await api.put(`/tasks/update/${taskId}`, data);
  validateResponse(response);
}