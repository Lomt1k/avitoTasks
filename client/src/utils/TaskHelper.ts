import { TaskPriority, TaskStatus } from "../api/Task";

export const getStatusLocalization = (taskStatus: string): string => {
  if (!Object.values(TaskStatus).includes(taskStatus as TaskStatus)) {
    return 'Неизвестно';
  }

  switch (taskStatus) {
    case TaskStatus.Backlog:
      return 'Нужно сделать';
    case TaskStatus.InProgress:
      return 'В работе';
    case TaskStatus.Done:
      return 'Выполнено';
    default:
      return 'Неизвестно';
  }
}

export const getPriorityLocalization = (priority: string): string => {
  if (!Object.values(TaskPriority).includes(priority as TaskPriority)) {
    return 'Неизвестно';
  }

  switch (priority) {
    case TaskPriority.Low:
      return 'Низкий';
    case TaskPriority.Medium:
      return 'Средний';
    case TaskPriority.High:
      return 'Высокий';
    default:
      return 'Неизвестно';
  }
}