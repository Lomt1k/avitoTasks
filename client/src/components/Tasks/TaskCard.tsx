import { FC, memo } from 'react';
import { Task } from '../../api/Task';
import { getPriorityLocalization, getStatusLocalization } from '../../utils/TaskHelper';
import './TaskCard.scss';

type TaskCardProps = {
  task: Task,
}

const TaskCard: FC<TaskCardProps> = memo(({ task }) => {
  const handleCardClick = () => {
    // TODO
  };

  return (
    <button className="task-card" onClick={handleCardClick} type="button">
      <div className="task-card__info">
        <h4 className="task-card__title">{task.title}</h4>
        <div className="task-card__stickers">
          <span className="task-card__sticker">{getPriorityLocalization(task.priority)}</span>
          <span className="task-card__sticker">{getStatusLocalization(task.status)}</span>
        </div>
        <p className="task-card__desc">{task.description}</p>
      </div>
      <div className="task-card__user">
        <img
          className="task-card__user-avatar"
          src={task.assignee.avatarUrl}
          alt={task.assignee.fullName}
          height={45}
        />
        <div className="task-card__user-info">
          <span className="task-card__user-name">{task.assignee.fullName}</span>
          <a
            href={`mailto:${task.assignee.email}`}
            className="task-card__user-email"
            onClick={(e) => e.stopPropagation()}
          >
            {task.assignee.email}
          </a>
        </div>
      </div>
    </button>
  );
});

export { TaskCard };