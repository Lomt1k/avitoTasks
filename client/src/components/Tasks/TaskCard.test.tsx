import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/react';
import { TaskCard } from './TaskCard';
import { Task } from '../../api/Task';
import { getPriorityLocalization, getStatusLocalization } from '../../utils/TaskHelper';

vi.mock('../CreateTaskModal', () => ({
  EditTaskModal: () => <span>Modal</span>,
}));

const task: Task = {
  assignee: {
    avatarUrl: 'http://example.com/avatar.jpg',
    email: 'user@example.com',
    fullName: 'User',
    id: 1,
  },
  boardId: 1,
  boardName: 'Board',
  description: 'Description',
  id: 1,
  priority: 'Low',
  status: 'Backlog',
  title: 'Task',
};

describe('TaskCard', () => {
  it('Рендерится корректно', () => {
    const { getByText, getByAltText } = render(<TaskCard task={task} />);

    expect(getByText('Task')).toBeInTheDocument();
    expect(getByText(getPriorityLocalization(task.priority))).toBeInTheDocument();
    expect(getByText(getStatusLocalization(task.status))).toBeInTheDocument();
    expect(getByText('Description')).toBeInTheDocument();
    expect(getByAltText('User')).toBeInTheDocument();
    expect(getByText('user@example.com')).toBeInTheDocument();
  });

  it('Обрабатывает событие клика', () => {
    const { getByText } = render(<TaskCard task={task} />);
    fireEvent.click(getByText('Task'));
    expect(getByText('Modal')).toBeInTheDocument();
  });

  it('Снапшот', () => {
    const { container } = render(<TaskCard task={task} />);
    expect(container).toMatchSnapshot();
  });
});
