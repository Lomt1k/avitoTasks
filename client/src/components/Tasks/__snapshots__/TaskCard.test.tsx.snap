// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard > Снапшот 1`] = `
<div>
  <button
    class="task-card"
    draggable="false"
    type="button"
  >
    <div
      class="task-card__info"
    >
      <h4
        class="task-card__title"
      >
        Task
      </h4>
      <div
        class="task-card__stickers"
      >
        <span
          class="task-card__sticker"
        >
          Низкий
        </span>
        <span
          class="task-card__sticker"
        >
          Нужно сделать
        </span>
      </div>
      <p
        class="task-card__desc"
      >
        Description
      </p>
    </div>
    <div
      class="task-card__user"
    >
      <img
        alt="User"
        class="task-card__user-avatar"
        height="45"
        src="http://example.com/avatar.jpg"
      />
      <div
        class="task-card__user-info"
      >
        <span
          class="task-card__user-name"
        >
          User
        </span>
        <a
          class="task-card__user-email"
          href="mailto:user@example.com"
        >
          user@example.com
        </a>
      </div>
    </div>
  </button>
</div>
`;
